grammar.js

```diff
@@ -110,17 +135,51 @@ module.exports = grammar({
       $.template_body
     ),
 
+    // The EBNF makes a distinction between function type parameters and other
+    // type parameters as you can't specify variance on function type
+    // parameters. This isn't important to the structure of the AST so we don't
+    // make that distinction.
     type_parameters: $ => seq(
       '[',
-      commaSep1($._type_parameter),
+      commaSep1($._variant_type_parameter),
       ']'
     ),
 
-    _type_parameter: $ => choice(
-      '_',
-      $.identifier
+    _variant_type_parameter: $ => seq(
+      choice(
+        $.covariant_type_parameter,
+        $.contravariant_type_parameter,
+        $._type_parameter // invariant type parameter
+      )
+    ),
+
+    covariant_type_parameter: $ => seq(
+      '+',
+      $._type_parameter
     ),
 
+    contravariant_type_parameter: $ => seq(
+      '-',
+      $._type_parameter,
+    ),
+
+    _type_parameter: $ => seq(
+      choice($.wildcard, $.identifier),
+      optional($.type_parameters),
+      optional($.upper_bound),
+      optional($.lower_bound),
+      optional(repeat($.view_bound)),
```

Comment:
The `repeat` rule actually means *zero-or-more* (there's also a `repeat1` built-in, which means *one-or-more*), so you can remove the `optional` here.